### PR TITLE
Add repositoryServerUserHostname to the kopia repository based functions

### PR DESCRIPTION
**Pull request by r4rajat**
## Change Overview

Add `repositoryServerUserHostname` as an Optional Argument to the Kopia Repository Based functions.

It enables to users to pass the hostname from the user access map which they want to use while performing kopia repository server based operations

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [x] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [x] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues 

- fixes #issue-number

## Test Plan

- [ ] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

#### 1) Create Images for Kanister and Repo Server controller

```
git tag -fa v21-repo-server-rajat -m "Testing"

bash build/gorelease.sh

docker tag ghcr.io/kanisterio/controller:v21-repo-server-rajat r4rajat/controller:v21-repo-server-rajat

docker tag ghcr.io/kanisterio/repo-server-controller:v21-repo-server-rajat r4rajat/repo-server-controller:v21-repo-server-rajat

docker push r4rajat/controller:v21-repo-server-rajat && docker push r4rajat/repo-server-controller:v21-repo-server-rajat
``` 

#### 2) Install Kanister 

```
helm install kanister ./helm/kanister-operator \
--namespace kanister \
--set image.repository=r4rajat/controller \
--set image.tag=v21-repo-server-rajat \
--set repositoryServerImage.repository=r4rajat/repo-server-controller \
--set repositoryServerImage.tag=v21-repo-server-rajat \ 
--set controller.parallelism=10  \
--create-namespace
```

#### 3) Apply Repo Server CRD

```
kubectl apply -f pkg/customresource/repositoryserver.yaml -n kanister
```

#### 4) Create Test Application [Time Logger]

```
kubectl create namespace time-logger

kubectl create -f ./examples/time-log/time-logger-deployment.yaml -n time-logger
```

#### 5) Create OpenSSL Certificate

```
openssl req -newkey rsa:2048 -nodes -keyout key.pem -x509 -days 365 -out certificate.pem
```

#### 6) Create S3 Location and Location Secret Config Files

- S3 Location Secret
```
vi s3_location_creds.yaml
```
```
apiVersion: v1
kind: Secret
metadata:
   name: s3-creds
   namespace: kanister
   labels:
      repo.kanister.io/target-namespace: monitoring
type: secrets.kanister.io/aws
data:
   # required: base64 encoded value for key with proper permissions for the bucket
   aws_access_key_id: <base64 encoded access key>
   # required: base64 encoded value for the secret corresponding to the key above
   aws_secret_access_key: <base64 encoded secret key>
```
- S3 Location
```
vi s3_location.yaml
```
```
apiVersion: v1
kind: Secret
metadata:
   name: s3-location
   namespace: kanister
   labels:
      repo.kanister.io/target-namespace: monitoring
type: Opaque
data:
   # required: specify the type of the store
   # supported values are s3, gcs, azure, and file-store
   type: czM=
   bucket: cmFqYXQtaW5mcmFjbG91ZA==
   # optional: used as a sub path in the bucket for all backups
   path: L3JlcG8tY29udHJvbGxlci8=
   # required, if supported by the provider
   region: dXMtZWFzdC0x
   # optional: if set to true, do not verify SSL cert.
   # Default, when omitted, is false
   #skipSSLVerify: false
   # required: if type is `file-store`
   # optional, otherwise
   #claimName: store-pvc
```

#### 7) Apply Secrets

```
kubectl create secret tls repository-server-tls-cert --cert=certificate.pem --key=key.pem -n kanister

kubectl create secret generic repository-server-user-access -n kanister --from-literal=localhost=test1234 --from-literal=localhost2=test12345 --from-literal=localhost3=test123456

kubectl create secret generic repository-admin-user -n kanister --from-literal=username=admin@testpod1 --from-literal=password=test1234

kubectl create secret generic repo-pass -n kanister --from-literal=repo-password=test1234

kubectl apply -f s3_location_creds.yaml -n kanister

kubectl apply -f s3_location.yaml -n kanister
```

#### 8) Create Repository

```
kopia --log-level=error --config-file=/tmp/kopia-repository.config --log-dir=/tmp/kopia-cache repository create --no-check-for-updates --cache-directory=/tmp/cache.dir --content-cache-size-mb=0 --metadata-cache-size-mb=500 --override-hostname=mysql.app --override-username=kanisterAdmin s3 --bucket=rajat-infracloud --prefix=/repo-controller/ --region=us-east-1 --access-key=<ACCESS_KEY> --secret-access-key=<SECRET_ACCESS_KEY>
```

#### 9) Create Repository Server CR

```
vi repo-server-cr.yaml
```

```
apiVersion: cr.kanister.io/v1alpha1
kind: RepositoryServer
metadata:
  labels:
    app.kubernetes.io/name: repositoryserver
    app.kubernetes.io/instance: repositoryserver-sample
    app.kubernetes.io/part-of: kanister
    app.kuberentes.io/managed-by: kustomize
    app.kubernetes.io/created-by: kanister
  name: kopia-repo-server-1
  namespace: kanister
spec:
  storage:
    secretRef:
      name: s3-location
      namespace: kanister
    credentialSecretRef:
      name: s3-creds
      namespace: kanister
  repository:
    rootPath: /repo-controller/
    passwordSecretRef:
      name: repo-pass
      namespace: kanister
    username: kanisterAdmin
    hostname: mysql.app
  server:
    adminSecretRef:
      name: repository-admin-user
      namespace: kanister
    tlsSecretRef:
      name: repository-server-tls-cert
      namespace: kanister
    userAccess:
      userAccessSecretRef:
        name: repository-server-user-access
        namespace: kanister
      username: kanisteruser
```

```bash
kubectl apply -f repo-server-cr.yaml -n kanister
```

Wait till the status of Repository Server CR gets to ```ServerReady``` , You could check it by running following command
```
kubectl describe -n kanister repositoryserver.cr.kanister.io/kopia-repo-server-1
```

#### 10) Create Blueprint
```
vi test-blueprint.yaml
```
```
apiVersion: cr.kanister.io/v1alpha1
kind: Blueprint
metadata:
  name: backupdate-bp
  namespace: kanister
actions:
  backup:
    outputArtifacts:
      timeLog:
        keyValue:
          path: '/repo-controller/time-logger/'
      backupIdentifier:
        keyValue:
          id: "{{ .Phases.backupToS3.Output.backupID }}"
    phases:
    - func: BackupDataUsingKopiaServer
      name: backupToS3
      args:
        namespace: "{{ .Deployment.Namespace }}"
        pod: "{{ index .Deployment.Pods 0 }}"
        container: test-container
        includePath: /var/log
        repositoryServerUserHostname: localhost2
```

```
kubectl create -f test-blueprint.yaml -n kanister
```

#### 11) Build kanctl with latest changes
```
go build -o kanctl cmd/kanctl/main.go 
```

#### 12) Take Backup of the Application
```
./kanctl create actionset --action backup --namespace kanister --blueprint backupdate-bp --deployment time-logger/time-logger --repository-server=kopia-repo-server-1

actionset backup-7m5lh created
```

Check Status of the actionset
```
kubectl describe actionsets -n kanister backup-7m5lh

Events:
  Type    Reason           Age   From                 Message
  ----    ------           ----  ----                 -------
  Normal  Started Action   9s    Kanister Controller  Executing action backup
  Normal  Started Phase    9s    Kanister Controller  Executing phase backupToS3
  Normal  Ended Phase      3s    Kanister Controller  Completed phase backupToS3
  Normal  Update Complete  3s    Kanister Controller  Updated ActionSet 'backup-7m5lh' Status->complete
```

#### 13) Update Blueprint with userHostname with value not in user access map
```
vi test-blueprint.yaml
```
```
apiVersion: cr.kanister.io/v1alpha1
kind: Blueprint
metadata:
  name: backupdate-bp
  namespace: kanister
actions:
  backup:
    outputArtifacts:
      timeLog:
        keyValue:
          path: '/repo-controller/time-logger/'
      backupIdentifier:
        keyValue:
          id: "{{ .Phases.backupToS3.Output.backupID }}"
    phases:
    - func: BackupDataUsingKopiaServer
      name: backupToS3
      args:
        namespace: "{{ .Deployment.Namespace }}"
        pod: "{{ index .Deployment.Pods 0 }}"
        container: test-container
        includePath: /var/log
        repositoryServerUserHostname: localhost20
```

```
kubectl apply -f test-blueprint.yaml -n kanister
```

#### 14) Take Backup of the Application
```
./kanctl create actionset --action backup --namespace kanister --blueprint backupdate-bp --deployment time-logger/time-logger --repository-server=kopia-repo-server-1

actionset backup-w85qd created
```

Check Status of the actionset
```
kubectl describe actionsets -n kanister backup-w85qd

Events:
  Type     Reason                          Age   From                 Message
  ----     ------                          ----  ----                 -------
  Normal   Started Action                  9s    Kanister Controller  Executing action backup
  Normal   Started Phase                   9s    Kanister Controller  Executing phase backupToS3
  Warning  ActionSetFailed Action: backup  9s    Kanister Controller  Failed to execute phase: v1alpha1.Phase{Name:"backupToS3", State:"pending", Output:map[string]interface {}(nil)}: Failed to fetch Hostname/User Passphrase from Secret: Failed to find Hostname in User Access Map: hostname provided in the repository server does not exist in the user access map
```

### CODE REVIEW

File: docs/functions.rst [hunk 1]
```diff
@@ -1607,6 +1607,7 @@ Arguments:
    `container`, Yes, `string`, name of the kanister sidecar container
    `includePath`, Yes, `string`, path of the data to be backed up
    `snapshotTags`, No, `string`, custom tags to be provided to the kopia snapshots
+   `userHostname`, No, `string`, user's hostname from user access map to be provided to the kopia repository server for performing backup
```
Comment (kale-amruta) [inline docs/functions.rst:1610]: We should change it to `ReposiotryServerUserHostname`

**Review by kale-amruta — COMMENTED** [review]
We should rename `userHostname`

**Review by viveksinghggits — COMMENTED** [review]
LGTM

File: docs/functions.rst (hunk 1 above)
Comment (r4rajat) [reply docs/functions.rst]: Done !

File: docs/functions.rst [hunk 2]
```diff
@@ -1607,6 +1607,7 @@ Arguments:
    `container`, Yes, `string`, name of the kanister sidecar container
    `includePath`, Yes, `string`, path of the data to be backed up
    `snapshotTags`, No, `string`, custom tags to be provided to the kopia snapshots
+   `repositoryServerUserHostname`, No, `string`, user's hostname to access the kopia repository server. Hostname would be available in the user access credential secret
```
Comment (viveksinghggits) [inline docs/functions.rst:1610]: does second sentence mean, user should figure out the hostname by looking at the user access credential secret?
What does it mean?

File: pkg/function/backup_data_using_kopia_server.go [hunk 3]
```diff
@@ -254,3 +267,11 @@ func userCredentialsAndServerTLS(tp *param.TemplateParams) (string, string, erro
 	}
 	return string(userCredJSON), string(certJSON), nil
 }
+
+func checkHostnameExistsInUserAccessMap(userAccessMap map[string]string, hostname string) error {
+	// check if hostname is provided in the repository server exists in the user access map
```
Comment (viveksinghggits) [inline pkg/function/backup_data_using_kopia_server.go:272]: this is the `hostname is provided` or it exists in the user access map?

File: pkg/function/backup_data_using_kopia_server.go [hunk 4]
```diff
@@ -254,3 +267,11 @@ func userCredentialsAndServerTLS(tp *param.TemplateParams) (string, string, erro
...
+	if _, ok := userAccessMap[hostname]; !ok {
+		return errors.New("hostname provided in the repository server does not exist in the user access map")
```
Comment (viveksinghggits) [inline pkg/function/backup_data_using_kopia_server.go:274]:
```suggestion
		return errors.New("hostname provided in the repository server CR does not exist in the user access map")
```
?

File: pkg/function/backup_data_using_kopia_server.go (hunk 3 above)
Comment (r4rajat) [reply pkg/function/backup_data_using_kopia_server.go:272]: multiple mapping of hostnames and passwords could be present in User Access Map inside Repository Server User Access Credentials Secret, So the end user has to provide one of the hostname which he/she wants to use for connecting with kopia repository server

File: docs/functions.rst (hunk 2 above)
Comment (r4rajat) [reply docs/functions.rst:1610]: Yes ! User could provide the hostname which he/she wants to use by looking at user access credentials secret. If it's not provided we'll get the last entry from the user access map which is present inside the secret by default.

File: pkg/function/backup_data_using_kopia_server.go [hunk 5]
```diff
@@ -223,18 +229,25 @@ func backupDataUsingKopiaServer(
...
 	for key, val := range userAccessMap {
 		hostName = key
 		userPassPhrase = val
 	}
+	if hostname != "" {
+		err := checkHostnameExistsInUserAccessMap(userAccessMap, hostname)
+		if err != nil {
+			return "", "", errors.Wrap(err, "Failed to find Hostname in User Access Map")
+		}
+		hostName = hostname
```
[the remark this answers is not in the document]
Comment (r4rajat) [reply pkg/function/backup_data_using_kopia_server.go:248]: Yes Vivek, so if we look at this function, we're taking `hostname` as an argument. So what we're doing is we're initially assigning the last value of hostname from user access map to `hostName`. That is the default case.

Later on we're checking if user has provided any `hostname` and if it exists in the user access map. If it's present, we're updating the `hostName`'s value with the `hostname` provided by the user and returning `hostName`

File: pkg/function/backup_data_using_kopia_server.go [hunk 6]
```diff
@@ -254,3 +267,11 @@ func userCredentialsAndServerTLS(tp *param.TemplateParams) (string, string, erro
 	}
 	return string(userCredJSON), string(certJSON), nil
 }
+
+func checkHostnameExistsInUserAccessMap(userAccessMap map[string]string, hostname string) error {
+	// check if hostname is provided by the user exists in the user access map
```
Comment (viveksinghggits) [inline pkg/function/backup_data_using_kopia_server.go:272]:
```suggestion
	// check if hostname that is provided by the user exists in the user access map
```

File: pkg/function/backup_data_using_kopia_server.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (viveksinghggits) [reply pkg/function/backup_data_using_kopia_server.go:248]: aah, I got confused between the casing of the variables, didn't see difference between `hostName` and `hostname`.
I don't think we need `hostName` right? Even if we just remove `hostName` I think we can work with `hostname` easily. Let me know if you have questions about this.

[the remark this answers is not in the document]
Comment (r4rajat) [reply pkg/function/backup_data_using_kopia_server.go]: Removed the `hostName` variable

File: pkg/function/backup_data_using_kopia_server.go [hunk 7]
```diff
@@ -223,24 +229,36 @@ func backupDataUsingKopiaServer(
 	return kopiacmd.ParseSnapshotCreateOutput(stdout, stderr)
 }
 
-func hostNameAndUserPassPhraseFromRepoServer(userCreds string) (string, string, error) {
+func hostNameAndUserPassPhraseFromRepoServer(userCreds, hostname string) (string, string, error) {
```
Comment (viveksinghggits) [inline pkg/function/backup_data_using_kopia_server.go:232]:
```suggestion
func hostNameAndUserPassPhraseFromRepoServer(userCreds, hostName string) (string, string, error) {
```

should `hostName` be preferred?

**Review by viveksinghggits — APPROVED** [review]

File: pkg/function/backup_data_using_kopia_server.go (hunk 7 above)
Comment (viveksinghggits) [reply pkg/function/backup_data_using_kopia_server.go:232]: I see `hostname` is used as directly or as substring in other var name. I think we should change them to `hostName` or `HostName`. Whatever applicable.

Comment (r4rajat) [reply pkg/function/backup_data_using_kopia_server.go:232]: Yeah, seems good. Will make the changes !

Comment (r4rajat) [reply pkg/function/backup_data_using_kopia_server.go]: Done !

**Comment by viveksinghggits** [comment]
Looks good to me, let's :ship: it.

**Pull request merged**